zerver/lib/queue.py

```diff
@@ -467,12 +467,19 @@ def queue_event_on_commit(queue_name: str, event: dict[str, Any]) -> None:
 
 
 def retry_event(
-    queue_name: str, event: dict[str, Any], failure_processor: Callable[[dict[str, Any]], None]
+    queue_name: str,
+    event: dict[str, Any],
+    failure_processor: Callable[[dict[str, Any]], None],
+    backoff_delay_seconds: int | None = None,
+    max_request_retries: int = MAX_REQUEST_RETRIES,
 ) -> None:
     if "failed_tries" not in event:
         event["failed_tries"] = 0
     event["failed_tries"] += 1
-    if event["failed_tries"] > MAX_REQUEST_RETRIES:
+    if event["failed_tries"] > max_request_retries:
         failure_processor(event)
     else:
+        if backoff_delay_seconds:
+            # TODO: This is wrong - the process can't sleep for so long.
+            time.sleep(backoff_delay_seconds)
```

Comment:
An approach that might kinda work would be to just re-insert the event at the end of the queue until `backoff_delay_seconds` has passed.... we'll still spin hot, though.

The actual fix for this likely involves reworking this specific queue to use `AbstractScheduledJob`. Can you start a #backend thread about how to do this? It might be we want to rework the queue to use that anyway in order to support delays for mobile notifications, and it's worth some discussion.

Reply:
https://github.com/zulip/zulip/pull/35882 is the issue for delayed mobile notifications that would share the refactoring work with this project.

(I'm not entirely sure we want to do the required investment at this time).